documentation/projects/proposals/dark_mode/20240325-implementation_plan_dark_mode.md

```diff
@@ -0,0 +1,263 @@
+
+![Openverse.org dark mode color palette](/_static/dark_mode_palette_example.png)
+
+While we _will_ include easy mechanisms for exceptions to this rule, they do not
+appear to be necessary based on the designs. Implementing our dark mode, then,
+should allow component authors to write components _once_, using semantic color
+names, that will automatically switch between their light and dark mode
+counterparts.
+
+### Implementation
+
+We will switch our color names defined in the tailwind configuration to use
+semantic names, for example replacing "pink" with "primary" and "yellow" with
+"complementary". Instead of hardcoding these colors in the Tailwind
+configuration, the tailwind configuration will reference CSS variables defined
+in our root css file. The value of the CSS variables will be switched based on a
+dark mode CSS class added to the HTML root when dark mode is enabled.
+
+Tailwind's built in `dark:` modifier can be used for any styles which need to
+override the default behavior or add dark-mode specific styles beyond the core
+palette swap.
+
+Here are pretend, simplified examples of this setup.
+
+In our primary CSS file, we setup two lists of CSS variables with different
+values when the `.dark-mode` class is present:
+
+```css
+:root {
+  --color-primary: black;
+  --color-secondary: white;
+}
+
+.dark-mode {
+  --color-primary: white;
+  --color-secondary: black;
+}
+```
+
+In our Tailwind config, we reference these variables:
+
+```js
+const config = {
+  theme: {
+    colors: {
+      primary: "var(--color-primary)",
+      secondary: "var(--color-secondary)",
+    },
+  },
+}
+```
+
+In a component, we use one Tailwind class to implement the correct color in
+light and dark modes:
+
+```html
+<template>
+  <!-- This will be black in light mode and white in dark mode -->
+  <p class="text-primary">Hello World</p>
+</template>
+```
+
+Finally, if we ever need an "escape hatch" to make sure, a component is, for
+example, _always_ black _regardless_ of dark mode:
+
+```html
+<template>
+  <!-- This will be black in light mode and black in dark mode -->
+  <p class="text-primary dark:text-secondary">Hello World</p>
+</template>
+```
+
+### Rejected alternative approach
+
+Using the `dark:` modifier _exclusively_ for dark mode styling. This is more
+explicit, but much, much more verbose, and would require extensive edits to
+every single component we have written. Instead of writing `bg-background`, for
+example, we would have to write `bg-white dark:bg-black` all throughout the
+codebase.
+
+## Expected Outcomes
+
+<!-- List any succinct expected products from this implementation plan. -->
+
+- Users writing components will have a "dark mode compatible by default"
+  experience. By default, they will not need to think much about dark mode.
+- Color names in the codebase will be replaced with semantic names.
+- Frontend developers will have easy tools to visually test components in light
+  and dark mode.
```

Comment:
Should this include expectations for visitors of the site? These are only devex outcomes. What about Openverse.org users?

```suggestion
- Users writing components will have a "dark mode compatible by default"
  experience. By default, they will not need to think much about dark mode.
- Color names in the codebase will be replaced with semantic names.
- Frontend developers will have easy tools to visually test components in light
  and dark mode.
- Visitors to the site will have a darkmode option available to them to toggle on/off
- Visitors with `prefers-color-scheme` set to dark will have dark mode automatically enabled
```

The toggle option I'm not sure is possible yet, because Francisco hasn't finished the designs for that interaction, but `prefers-color-scheme` is a particularly important one to consider seeing as we're using a custom selector to indicate when dark mode should be used.

Reply:
Perfect, I'll commit this as-is.

Reply:
Added in 6295bb0a496428a19e22329f4dd4e1c863a60f17.